flows/vision-zero/cris_import/cris_import.py

```diff
@@ -546,59 +611,186 @@ def align_records(typed_token, dry_run):
 
     # fmt: on
-    return True
+    return map_state
+
+
+@task(
+    name="Group CSVs into logical groups",
+)
+def group_csvs_into_logical_groups(extracted_archives, dry_run):
+    files = os.listdir(str(extracted_archives))
+    logical_groups = []
+    for file in files:
+        if file.endswith(".xml"):
+            continue
+        match = re.search("^extract_(\d+_\d+)_", file)
+        group_id = match.group(1)
+        if group_id not in logical_groups:
+            logical_groups.append(group_id)
+    print("logical groups: " + str(logical_groups))
+    map_safe_state = []
+    for group in logical_groups:
+        map_safe_state.append({
+            "logical_group_id": group,
+            "working_directory": str(extracted_archives),
+            "csv_prefix": "extract_" + group + "_",
+            "dry_run": dry_run,
+        })
+    print(map_safe_state)
+    return map_safe_state
+
+
+@task(
+    name="Generate a short alphanumeric string based on logical group id",
+)
+def create_import_schema_name(mapped_state):
+    print(mapped_state)
+    schema = 'import_' + hashlib.md5(mapped_state["logical_group_id"].encode()).hexdigest()[:12]
+    mapped_state["import_schema"] = schema
+    print("Schema name: ", mapped_state["import_schema"])
+    return mapped_state
 
 
+@task(
+    name="Create target import schema",
+)
+def create_target_import_schema(map_state):
+    ssh_tunnel = SSHTunnelForwarder(
+        (DB_BASTION_HOST),
+        ssh_username=DB_BASTION_HOST_SSH_USERNAME,
+        ssh_private_key= '/root/.ssh/id_rsa', # will switch to ed25519 when we rebuild this for prefect 2
+        remote_bind_address=(DB_RDS_HOST, 5432)
+        )
+    ssh_tunnel.start()   
+
+    pg = psycopg2.connect(
+        host='localhost', 
+        port=ssh_tunnel.local_bind_port,
+        user=DB_USER, 
+        password=DB_PASS, 
+        dbname=DB_NAME, 
+        sslmode=DB_SSL_REQUIREMENT, 
+        sslrootcert="/root/rds-combined-ca-bundle.pem"
+        )
+
+    cursor = pg.cursor()
+    
+    # check if the schema exists by querying the pg_namespace system catalog
+    cursor.execute(f"SELECT EXISTS (SELECT 1 FROM pg_namespace WHERE nspname = '{map_state['import_schema']}')")
+
+    schema_exists = cursor.fetchone()[0]
+
+    # if the schema doesn't exist, create it using a try-except block to handle the case where it already exists
+    if not schema_exists:
+        try:
+            cursor.execute(f"CREATE SCHEMA {map_state['schema_name']}")
+            print("Schema created successfully")
+        except psycopg2.Error as e:
+            print(f"Error creating schema: {e}")
+    else:
+        print("Schema already exists")
+
+    # commit the changes and close the cursor and connection
+    pg.commit()
+    cursor.close()
+    pg.close()
+
+    return map_state
+
+@task(
+    name="Clean up import schema",
+)
+def clean_up_import_schema(map_state):
```

Comment:
Because we're creating a schema per import to stay thread safe, we need to clean them up.